src/libraries/System.IO.Compression/src/System/IO/Compression/WinZipAesStream.cs

```diff
@@ -0,0 +1,744 @@
+        }
+
+        private Task WriteHeaderAsync(CancellationToken cancellationToken)
+        {
+            return WriteHeaderCoreAsync(isAsync: true, cancellationToken);
+        }
+
+        private void ProcessBlock(byte[] buffer, int offset, int count)
+        {
+            Debug.Assert(_aesEncryptor is not null, "Cipher should have been initialized before processing blocks");
+
+            int processed = 0;
+
+            while (processed < count)
+            {
+                // Ensure we have enough keystream bytes available
+                int keystreamAvailable = KEYSTREAM_BUFFER_SIZE - _keystreamOffset;
+                if (keystreamAvailable == 0)
+                {
+                    GenerateKeystreamBuffer();
+                    keystreamAvailable = KEYSTREAM_BUFFER_SIZE;
+                }
+
+                // Process as many bytes as possible with the available keystream
+                int bytesToProcess = Math.Min(count - processed, keystreamAvailable);
+
+                // For encryption: XOR first, then HMAC the ciphertext
+                if (_encrypting)
+                {
+                    // XOR the data with the keystream to create ciphertext
+                    XorBytes(buffer, offset + processed, _keystreamBuffer, _keystreamOffset, bytesToProcess);
+                    // HMAC is computed on the ciphertext (after XOR)
+                    _hmac.TransformBlock(buffer, offset + processed, bytesToProcess, null, 0);
+                }
+                // For decryption: HMAC first (on ciphertext), then XOR
+                else
+                {
+                    // HMAC is computed on the ciphertext (before XOR)
+                    _hmac.TransformBlock(buffer, offset + processed, bytesToProcess, null, 0);
+                    // XOR the ciphertext with the keystream to recover plaintext
+                    XorBytes(buffer, offset + processed, _keystreamBuffer, _keystreamOffset, bytesToProcess);
+                }
+
+                _keystreamOffset += bytesToProcess;
+                processed += bytesToProcess;
+            }
+        }
+
+        private void GenerateKeystreamBuffer()
+        {
+            Debug.Assert(_aesEncryptor is not null, "Cipher should have been initialized");
+
+            // Generate KEYSTREAM_BUFFER_SIZE bytes of keystream (256 blocks of 16 bytes each)
+            for (int i = 0; i < KEYSTREAM_BUFFER_SIZE; i += BLOCK_SIZE)
+            {
+                _aesEncryptor.TransformBlock(_counterBlock, 0, BLOCK_SIZE, _keystreamBuffer, i);
+                IncrementCounter();
+            }
+
+            _keystreamOffset = 0;
+        }
+
+        private static void XorBytes(byte[] dest, int destOffset, byte[] src, int srcOffset, int count)
+        {
+            Span<byte> destSpan = dest.AsSpan(destOffset, count);
+            ReadOnlySpan<byte> srcSpan = src.AsSpan(srcOffset, count);
```

Comment:
Since the very first thing you do is transform the array arguments into spans, why not accept those spans as arguments directly?